spec/unit/parser_spec.rb

```diff
@@ -0,0 +1,255 @@
+    attr_reader :params
+    def initialize(params)
+      @params = params
+    end
+  end
+
+  before { Controller.include IknowParams::Parser }
+
+  describe "#parse_param" do
+
+    let(:parsed) { Controller.new(@params).parse_param(@param, with: serializer) }
+
+    context "no serializer specified" do
+      let(:serializer) { nil }
+
+      it "returns the value" do
+        @params = {name: "foo"}
+        @param  = :name
+
+        expect(parsed).to eq @params[@param]
+      end
+
+      it "requires the value" do
+        @params = {id: 5}
+        @param  = :name
+
+        expect { parsed }.to raise_error(IknowParams::Parser::ParseError)
+      end
+
+      context "with a default specified" do
+        let(:default) { Object.new }
+        let(:parsed) { Controller.new(@params).parse_param(@param, with: serializer, default: default) }
+
+        it "returns the value when param is present" do
+          @params = {name: "foo"}
+          @param  = :name
+
+          expect(parsed).to eq @params[@param]
+        end
+
+        it "returns the default when param is missing" do
+          @params = {id: 5}
+          @param  = :name
+
+          expect(parsed).to eq default
+        end
+      end
+
+      context "when dumping the value" do
+        let(:parsed) { Controller.new(@params).parse_param(@param, with: serializer, dump: true) }
+
+        it "raises an error" do
+          @params = {name: "foo"}
+          @param  = :name
+
+          expect { parsed }.to raise_error(IknowParams::Parser::ParseError)
+        end
+      end
+    end
+
+    context "with a serializer specified" do
+      let(:serializer) { :Integer }
+
+      it "returns the parsed value" do
+        @params = {id: "5"}
+        @param  = :id
+
+        expect(parsed).to eq 5
+      end
+
+      it "requires the value" do
+        @params = {name: "5"}
+        @param  = :id
+
+        expect { parsed }.to raise_error(IknowParams::Parser::ParseError)
+      end
+
+      it "raises an error when the param can't be parsed into the serializer's type" do
+        @params = {id: "fish"}
+        @param  = :id
+
+        expect { parsed }.to raise_error(IknowParams::Parser::ParseError)
+      end
+
+      context "with a default specified" do
+        let(:default) { 9000 }
+        let(:parsed) { Controller.new(@params).parse_param(@param, with: serializer, default: default) }
+
+        it "returns the value when param is present" do
+          @params = {id: "10"}
+          @param  = :id
+
+          expect(parsed).to eq 10
+        end
+
+        it "returns the default when param is missing" do
+          @params = {name: "22"}
+          @param  = :id
+
+          expect(parsed).to eq default
+        end
+      end
+
+      context "when dumping the value" do
+        let(:parsed) { Controller.new(@params).parse_param(@param, with: serializer, dump: true) }
+
+        it "returns the string representation" do
+          @params = {id: "10"}
```

Comment:
Consider testing the normalization by reserializing, by for example parsing `0xa` and dumping `10`? (Also, gosh `Integer()` has an entertainingly extensive parser)